presto-matching/src/main/java/com/facebook/presto/matching/Matcher.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.matching;
+
+import com.facebook.presto.matching.pattern.CapturePattern;
+import com.facebook.presto.matching.pattern.TypeOfPattern;
+import com.facebook.presto.matching.pattern.WithPattern;
+
+public interface Matcher
+{
+    default <T> Match<T> match(Pattern<T> pattern, Object object)
+    {
+        return match(pattern, object, Captures.empty());
+    }
+
+    <T> Match<T> match(Pattern<T> pattern, Object object, Captures captures);
```

Comment:
I had a broader concept in mind of a 'pattern `Interpreter`'. `Matcher` would be one of the `Interpreter`s. The interpreters would differ in return type*. E.g.:
 - Matchers would return a Match
 - MultiMatchers would return all possible matches whenever there's an 'alternative' pattern involved - so they'd return a `List<Match<T>>` or a Stream
 - Builders would build an object instance matching the given pattern (filling the gaps with customizable defaults)
 - Generators would do the same, but use random data and return a Stream of objects - great for property based testing
 - Compilers would return a method handle to an optimized `match` method, or rather a compilation context

The `match` method would be the entrypoint for a `Matcher`. The `evaluate` methods (or rather: `interpret`) would structure the interpretation of a pattern (same for all interpreters)

I guess I'll name all the methods `match` for now. But since you asked, hope you had a good time reading this :)

`*` I'm not yet sure how to bend the Java type system to achieve different return types in a generic way.